### PR TITLE
[numpy] Infra for supporting numpy ops in imperative mode and Gluon APIs

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
This PR adds necessary infrastructural components to support numpy operators in imperative programming and Gluon APIs.
1. Namespace `mxnet.numpy` is used for registering numpy ops used in imperative programming as how people use the official NumPy package.
2. Namepsaces `mxnet.ndarray.numpy` and `mxnet.symbol.numpy` are for registering numpy ops used in Gluon APIs dispatched by `F`. Users, for example, can use `np.sum` in Gluon as `F.np.sum`.
3. A new `ndarray` class that is visible to users with the same convenience fluent methods as in official `numpy.ndarray`. Currently, it only supports data layout of c-contiguous. Strides and other storage types will be supported in the future.
4. A new `_NumpySymbol` class that is transparent to users. This type of objects are created automatically by the underlying mechanism in Gluon `HybridBlock` for building computational graphs after `hybridize()` is called. It provides the same set of convenience fluent methods as `ndarray` and users should not interact with it directly.
5. All numpy ops generate tensors of type `ndarray`, instead of the classic `NDArray`. This includes both imperative_invoke and CachedOp. In the situations where CachedOp is used, the output tensors' types are determined by the immediate operators that create them. If an output is generated by a numpy op, it is of type `ndarray`; if generated by a legacy op, it is of type `NDArray`.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/numpy/multiarray.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    def sin(self, *args, **kwargs):
+        """Convenience fluent method for :py:func:`sin`.
+
+        The arguments are the same as for :py:func:`sin`, with
+        this array as data.
+        """
+        raise NotImplementedError('mxnet.numpy.ndarray.sin is not implemented. Please '
+                                  'convert the mxnet.numpy.ndarray to mxnet.ndarray.NDArray '
+                                  'and call the sin function as follows: '
+                                  'self.as_legacy_ndarray().sin(*args, **kwargs).')
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: Maybe refine the error message to be more clear that numpy actually does not have convenient method for basic mathematical calculations? Same for the other convenient ones that already exists.

File: include/mxnet/c_api.h [hunk 2]
```diff
@@ -2794,6 +2794,36 @@ MXNET_DLL int MXEnginePushSync(EngineSyncFunc sync_func, void* func_param,
...
+ */
+MXNET_DLL int MXShallowCopyNDArray(NDArrayHandle src, NDArrayHandle* out);
+/*!
+ * \brief Create an Symbol from source sharing the same graph structure.
+ * \param src source Symbol
+ * \param out new Symbol sharing the same graph structure with src
+ */
+MXNET_DLL int MXShallowCopySymbol(SymbolHandle src, SymbolHandle * out);
+/*!
+ * \brief Create an Symbol from source sharing the same graph structure.
```
Comment (szha) [inline include/mxnet/c_api.h]: update doc

File: python/mxnet/_ctypes/ndarray.py [hunk 3]
```diff
@@ -93,13 +101,28 @@ def _imperative_invoke(handle, ndargs, keys, vals, out):
 
     if original_output is not None:
         return original_output
-    if num_output.value == 1:
-        return _ndarray_cls(ctypes.cast(output_vars[0], NDArrayHandle),
-                            stype=out_stypes[0])
+    if _is_np_op(handle):
```
Comment (szha) [inline python/mxnet/_ctypes/ndarray.py]: to simplify:
```python
create_array_fn = _np_ndarray_cls if _is_np_op(handle) else _ndarray_cls
if num_output.value == 1:
    return create_array_fn(ctypes.cast(output_vars[0], NDArrayHandle),
                           stype=out_stypes[0])
...
```

File: python/mxnet/_ctypes/ndarray.py [hunk 4]
```diff
@@ -152,9 +175,12 @@ def __call__(self, *args, **kwargs):
         if original_output is not None:
             return original_output
         if num_output.value == 1:
-            return _ndarray_cls(ctypes.cast(output_vars[0], NDArrayHandle),
-                                stype=out_stypes[0])
+            if _is_output_from_np_op(self.handle, 0):
```
Comment (szha) [inline python/mxnet/_ctypes/ndarray.py]: same as above

File: python/mxnet/_ctypes/ndarray.py [hunk 5]
```diff
@@ -93,13 +101,28 @@ def _imperative_invoke(handle, ndargs, keys, vals, out):
...
+            return _ndarray_cls(ctypes.cast(output_vars[0], NDArrayHandle),
+                                stype=out_stypes[0])
+        else:
+            return [_ndarray_cls(ctypes.cast(output_vars[i], NDArrayHandle),
+                                 stype=out_stypes[i])
+                    for i in range(num_output.value)]
+
+
+def _is_output_from_np_op(handle, idx):
+    """Check if the CachedOp's idx-th output is from a numpy op."""
```
Comment (szha) [inline python/mxnet/_ctypes/ndarray.py]: if this is only for cached op, should it be made part of CachedOp class?

File: include/mxnet/c_api.h (hunk 2 above)
Comment (reminisce) [reply include/mxnet/c_api.h]: Oops, forgot to change to the docstring. :)

File: python/mxnet/_ctypes/ndarray.py (hunk 3 above)
Comment (reminisce) [reply python/mxnet/_ctypes/ndarray.py]: Done

File: python/mxnet/_ctypes/ndarray.py (hunk 5 above)
Comment (reminisce) [reply python/mxnet/_ctypes/ndarray.py]: Done.

File: python/mxnet/_ctypes/ndarray.py (hunk 4 above)
Comment (reminisce) [reply python/mxnet/_ctypes/ndarray.py]: Done

File: include/mxnet/c_api.h (hunk 2 above)
Comment (reminisce) [reply include/mxnet/c_api.h]: Done

**Comment by reminisce** [comment]
@szha I addressed all your cr comments and added detailed descriptions in the beginning of this PR. I am adding unit tests. Thanks for the review.

**Pull request merged**